src/vos/vos_tree.c

```diff
@@ -977,12 +977,12 @@ key_tree_punch(struct vos_object *obj, daos_handle_t toh, daos_epoch_t epoch,
 	daos_handle_t		 loh = DAOS_HDL_INVAL;
 	int			 rc;
 
-	rc = dbtree_fetch(toh, BTR_PROBE_EQ, DAOS_INTENT_UPDATE, key_iov, NULL,
+	rc = dbtree_fetch(toh, BTR_PROBE_EQ, DAOS_INTENT_PUNCH, key_iov, NULL,
 			  val_iov);
 	if (rc != 0) {
 		D_ASSERT(rc == -DER_NONEXIST);
 		/* use BTR_PROBE_BYPASS to avoid probe again */
-		rc = dbtree_upsert(toh, BTR_PROBE_BYPASS, DAOS_INTENT_UPDATE,
+		rc = dbtree_upsert(toh, BTR_PROBE_BYPASS, DAOS_INTENT_PUNCH,
```

Comment:
this should also have no effect.   We don't do anything with intents in the key/object tree anymore.

Reply:
Given "UPDATE" intent for punch logic looks confused, although it does not affect the final correctness, that depends on incarnation log logic. But at current point, dbtree_upsert() does not know how the "intent" will be used by lower layer logic, it will be more clear if we offer the exact "intent".

Reply:
My point is we could just pass 0.  The intent isn't used.

Reply:
Actually, the more I think about it, passing 0 would be best.   It would make it clear that it's not important what we pass there.

Reply:
Passing "0" will establish strong interconnection with incarnation log logic (hidden behind dbtree), that assumes the "intent" parameter will only be used (and ignored) by incarnation log logic. If we adjust dbtree logic (to use the "intent") or incarnation log logic someday, we may have to correct this "intent". So offering the exact "intent" is safe and better.

Reply:
If the intent doesn't take effect here, I agree with Jeff that we'd pass DAOS_INTENT_DEFAULT like other places where the common dbtree API is used (such as VEA, SMD, etc.), that looks clearer to me

I'm wondering that could be even better if we define 0 as "DAOS_INTENT_NONE" (all other valid intents starts from 1) and use it in such cases, but it requires much more code changes.

Reply:
The "intent" is useless for incarnation log, can be regarded as special case. But from current caller perspective, it calls dbtree API. We should NOT assume the dbtree logic will not use the "intent" internally. In this special case, the dbtree API pass the "intent" parameter to lower layer API, here it is incarnation log. But the lower layer can be others, just like the case before incarnation log introduced (at that time, the "intent" is necessary).

On the other hand, the "intent" parameter for dbtree API is still useful for single value tree, so we cannot remove "intent' now.

Reply:
Right, but in the case of key_tree_punch, intent is meaningless for the dbtree.    It's not used internally.   The dbtree logic only fetches the key, the incarnation log is external to this call entirely (see the ilog_update call a few lines below).   In this case, it passes "true" as the punch parameter.   The intent argument here does nothing.   If anything, I think we could do your flag idea but I don't see how this clarifies anything in this case.   Niu's suggestion is also reasonable...something like DAOS_INTENT_NONE or DAOS_INTENT_IGNORED would make it clearer.   Or using flags and passing 0.

Reply:
Just realize that we can remove "intent" parameter from dbtree_upsert() API, because we have another patch to NOT create (to be punched) target if it does non exist. Then all the callers for dbtree_upsert() can use the same intent by default. I will close this PR, and the change will be made in the patch PR#1856